File: src/engine/src/piece_queue.rs
/*
 * Copyright (c) Grigory Alfyorov. Licensed under the MIT License.
 * See the LICENCE file in the repository root for full licence text.
 */

use std::collections::VecDeque;
use crate::piece::PieceType;
use crate::piece_generators::{PieceGenerator, PieceGeneratorBag7};

#[derive(Debug)]
pub struct PieceQueue {
    pub queue: VecDeque<PieceType>,
    piece_generator: Box<PieceGeneratorBag7>,
    next_piece: PieceType,
    seed: u64
}

impl PieceQueue {
    pub fn new(seed: u64) -> Self {

        let mut piece_generator = Box::new(PieceGeneratorBag7::new(seed));
        let queue = piece_generator.init();

        Self {
            queue,
            piece_generator,
            next_piece: PieceType::Pixel,
            seed
        }
    }

    pub fn next(&mut self) -> PieceType {
        self.set_piece()
    }

    fn set_piece(&mut self) -> PieceType {
        self.next_piece = self.piece_generator.next();
        let next = self.queue.pop_front().unwrap();

        self.queue.push_back(self.next_piece);

        next
    }

    pub fn reset(&mut self, new_seed: Option<u64>) {

        let seed = new_seed.unwrap_or_else(|| self.seed);

        let mut piece_generator = Box::new(PieceGeneratorBag7::new(seed));
        let queue = piece_generator.init();

        self.queue = queue;
        self.piece_generator = piece_generator;
        self.next_piece = PieceType::Pixel;
        self.seed = seed;
    }
}

File: csharp/Quader/Quader.Engine/Pieces/PieceQueue.cs
using System.Collections;
using Quader.Engine.Pieces.Generators;

namespace Quader.Engine.Pieces;

public class PieceQueue : IEnumerable<PieceType>
{
    private Queue<PieceType> _queue;
    private IPieceGenerator _pieceGenerator;
    private PieceType _nextPiece;
    private int _seed;

    public PieceQueue(int seed)
    {
        _seed = seed;
        _pieceGenerator = new PieceGeneratorBag7(_seed);
        _nextPiece = PieceType.Pixel;

        var q = _pieceGenerator.Initialize();
        _queue = new Queue<PieceType>();
        foreach (var pieceType in q)
            _queue.Enqueue(pieceType);
    }

    public PieceType NextPiece() => SetPiece();

    public void Reset(int? newSeed)
    {
        if (newSeed.HasValue) _seed = newSeed.Value;

        var pieceGenerator = new PieceGeneratorBag7(_seed);
        var queue = pieceGenerator.Initialize();

        _queue = new Queue<PieceType>();
        foreach (var pieceType in queue)
            _queue.Enqueue(pieceType);

        _pieceGenerator = pieceGenerator;
        _nextPiece = PieceType.Pixel;
    }

    private PieceType SetPiece()
    {
        _nextPiece = _pieceGenerator.Next();
        var next = _queue.Dequeue();

        _queue.Enqueue(_nextPiece);

        return next;
    }

    public IEnumerator<PieceType> GetEnumerator()
    {
        return _queue.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}